Applications/SpaceAnalyzer/TreeMapWidget.h

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <LibGUI/Frame.h>
+#include <LibGfx/Rect.h>
+
+namespace TreeMap {
+
+struct ITreeMapNode {
+    virtual String name() = 0;
+    virtual int64_t area() = 0;
+    virtual size_t num_children() = 0;
+    virtual ITreeMapNode& child_at(size_t i) = 0;
+    virtual void sort_children_by_area() = 0;
+};
+
+struct ITreeMap : public RefCounted<ITreeMap> {
+    virtual ~ITreeMap() { }
+    virtual ITreeMapNode* get_root() = 0;
+};
+
+class TreeMapWidget final : public GUI::Frame {
+    C_OBJECT(TreeMapWidget)
+public:
+    virtual ~TreeMapWidget() override;
+    Function<void()> on_path_change;
+    size_t path_size();
+    ITreeMapNode* get_nth_path_node(size_t n);
+    size_t get_viewpoint();
+    void set_viewpoint(size_t);
+    void set_tree(RefPtr<ITreeMap> tree);
+
+private:
+    TreeMapWidget();
+    virtual void paint_event(GUI::PaintEvent&) override;
+    virtual void mousedown_event(GUI::MouseEvent&) override;
+    virtual void doubleclick_event(GUI::MouseEvent&) override;
+    virtual void mousewheel_event(GUI::MouseEvent&) override;
+
+    bool area_can_contain_children(Gfx::IntRect ext);
+    bool area_can_contain_label(Gfx::IntRect ext);
```

Comment:
These should be `const` qualified. Also please pass large argument types such as `Gfx::IntRect` by const reference.